src/ZODB/Connection.py

```diff
@@ -101,11 +103,10 @@ def __init__(self, db, cache_size=400, before=None, cache_size_bytes=0):
         # Multi-database support
         self.connections = {self._db.database_name: self}
 
-        storage = db._mvcc_storage
         if before:
-            storage = storage.before_instance(before)
+            storage = HistoricalStorageAdapter(db.storage, before)
```

Comment:
I'm slightly concerned too, but the test passes.

I thought of the lack of new_instance after I pushed this. Thinking about it, I suspect it would break with multiple threads.

Your change (which is what I used initially) doesn't work unless I add loadBefore to the MVCCAdapterInstance, which I was trying to avoid.  I'll go ahead and add it for now.